### PR TITLE
[dynamo][guards] Use dict for storing weakrefs

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107496
* __->__ #107645

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov

### CODE REVIEW

**Comment by voznesenskym** [comment]
Why?

**Comment by anijain2305** [comment]
I am going to need to look into weakrefs from inside the GuardBuilder in my next PR. Also, it does not make sense to me why have 2 different data structure, when they could be shared into one.

File: torch/_dynamo/guards.py [hunk 1]
```diff
@@ -928,7 +927,6 @@ def source_ref(source):
         self.check_fn = self.compile_check_fn(
             local_builder, global_builder, guards, guard_fail_fn
         )
-        self._seen_ids.clear()
```
Comment (ezyang) [inline torch/_dynamo/guards.py:931]: `_weakrefs` doesn't seem to get cleared here?

**Review by ezyang — APPROVED** [review]

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**